### PR TITLE
Headless: Add support for fullscreen option

### PR DESCRIPTION
**Pull request by theundebruijn**
The headless SDL2 client was missing a way to launch in fullscreen mode.
This PR adds an optional argument similar to the one used in the GUI client's CLI.

In addition to getting headless in line with Ava/GTK, by adding this we support launching games in fullscreen mode from third party launchers (e.g. Steam Big Picture).

---

NOTE: This PR serves as a prerequisite for https://github.com/Ryujinx/Ryujinx/pull/5412

---

#### TODO's

_Implementation_ :
- [x] Implement `borderless fullscreen windowed` logic
- [x] Add `--fullscreen` option to Headless CLI argument list

_Testing_ :
- [x] Windows 11 (22H2) (Intel x64) — Nvidia (Vulkan + OpenGL)
- [x] MacOS 12.x (Intel x64) — AMD (Vulkan)
- [x] MacOS 13.x (ARM64) — Apple Silicon (Vulkan) (tyvm @siegmund-heiss-ich)
- [x] Linux (Fedora Workstation 38) (Intel x64) (Wayland) — AMD (Vulkan + OpenGL)
- [x] Linux (Fedora Workstation 38) (Intel x64) (X11) — AMD (Vulkan + OpenGL)

### CODE REVIEW

File: src/Ryujinx.Headless.SDL2/Program.cs [hunk 1]
```diff
@@ -439,6 +440,11 @@ void LoadPlayerConfiguration(string inputProfileName, string inputId, PlayerInde
             GraphicsConfig.ShadersDumpPath = option.GraphicsShadersDumpPath;
             GraphicsConfig.EnableMacroHLE = !option.DisableMacroHLE;
 
+            if (option.StartFullscreen)
+            {
+                StartFullscreen = true;
```
Comment (patrick-hovsepian) [inline src/Ryujinx.Headless.SDL2/Program.cs:445]: super nit: drop the condition and just assign to option.StartFullScreen

File: src/Ryujinx.Headless.SDL2/WindowBase.cs [hunk 2]
```diff
@@ -211,6 +211,14 @@ private void HandleWindowEvent(SDL_Event evnt)
 
         protected abstract void SwapBuffers();
 
+        private SDL_WindowFlags GetFullscreenFlags()
+        {
+            // NOTE: We default to 'borderless fullscreen windowed' mode.
+            // There's a case to be made to support a 'real' fullscreen mode. This would allow for things like Variable Refresh Rate support.
+            // Doing so would require additional cli options to pass in resolution parameters.
+            return (Program.StartFullscreen) ? SDL_WindowFlags.SDL_WINDOW_FULLSCREEN_DESKTOP : 0;
```
Comment (patrick-hovsepian) [inline src/Ryujinx.Headless.SDL2/WindowBase.cs:219]: is there a defined flag for 0?

File: src/Ryujinx.Headless.SDL2/Options.cs [hunk 3]
```diff
@@ -14,6 +14,9 @@ public class Options
         [Option("profile", Required = false, HelpText = "Set the user profile to launch the game with.")]
         public string UserProfile { get; set; }
 
+        [Option("fullscreen", Required = false, HelpText = "Launch the game in fullscreen mode.")]
+        public bool StartFullscreen { get; set; }
```
Comment (AcK77) [inline src/Ryujinx.Headless.SDL2/Options.cs:18]:
```suggestion
        public bool IsFullscreen { get; set; }
```

File: src/Ryujinx.Headless.SDL2/Program.cs [hunk 4]
```diff
@@ -439,6 +440,8 @@ void LoadPlayerConfiguration(string inputProfileName, string inputId, PlayerInde
             GraphicsConfig.ShadersDumpPath = option.GraphicsShadersDumpPath;
             GraphicsConfig.EnableMacroHLE = !option.DisableMacroHLE;
 
+            StartFullscreen = option.StartFullscreen;
```
Comment (AcK77) [inline src/Ryujinx.Headless.SDL2/Program.cs:443]: Why do not set a field in `WindowBase` instead ?

```suggestion
            _window.IsFullscreen = option.IsFullscreen;
```

File: src/Ryujinx.Headless.SDL2/WindowBase.cs [hunk 5]
```diff
@@ -156,7 +156,7 @@ private void InitializeWindow()
             string titleIdSection = string.IsNullOrWhiteSpace(activeProcess.ProgramIdText) ? string.Empty : $" ({activeProcess.ProgramIdText.ToUpper()})";
             string titleArchSection = activeProcess.Is64Bit ? " (64-bit)" : " (32-bit)";
 
-            WindowHandle = SDL_CreateWindow($"Ryujinx {Program.Version}{titleNameSection}{titleVersionSection}{titleIdSection}{titleArchSection}", SDL_WINDOWPOS_UNDEFINED, SDL_WINDOWPOS_UNDEFINED, DefaultWidth, DefaultHeight, DefaultFlags | GetWindowFlags());
+            WindowHandle = SDL_CreateWindow($"Ryujinx {Program.Version}{titleNameSection}{titleVersionSection}{titleIdSection}{titleArchSection}", SDL_WINDOWPOS_UNDEFINED, SDL_WINDOWPOS_UNDEFINED, DefaultWidth, DefaultHeight, DefaultFlags | GetFullscreenFlags() | GetWindowFlags());
```
Comment (AcK77) [inline src/Ryujinx.Headless.SDL2/WindowBase.cs:159]:
```suggestion
            SDL_WindowFlags fullscreenFlag =  IsFullscreen ? SDL_WindowFlags.SDL_WINDOW_FULLSCREEN_DESKTOP : 0;

            WindowHandle = SDL_CreateWindow($"Ryujinx {Program.Version}{titleNameSection}{titleVersionSection}{titleIdSection}{titleArchSection}", SDL_WINDOWPOS_UNDEFINED, SDL_WINDOWPOS_UNDEFINED, DefaultWidth, DefaultHeight, DefaultFlags | fullscreenFlag | GetWindowFlags());
```

File: src/Ryujinx.Headless.SDL2/WindowBase.cs [hunk 6]
```diff
@@ -211,6 +211,14 @@ private void HandleWindowEvent(SDL_Event evnt)
 
         protected abstract void SwapBuffers();
 
+        private SDL_WindowFlags GetFullscreenFlags()
```
Comment (AcK77) [inline src/Ryujinx.Headless.SDL2/WindowBase.cs:214]: This could be removed since it will not be used elsewhere

**Review by AcK77 — CHANGES_REQUESTED** [review]

File: src/Ryujinx.Headless.SDL2/Program.cs [hunk 7]
```diff
@@ -43,6 +43,7 @@ namespace Ryujinx.Headless.SDL2
     class Program
     {
         public static string Version { get; private set; }
+        public static bool StartFullscreen { get; private set; } = false;
```
Comment (AcK77) [inline src/Ryujinx.Headless.SDL2/Program.cs:46]: This is not needed anymore

File: src/Ryujinx.Headless.SDL2/WindowBase.cs [hunk 8]
```diff
@@ -156,7 +157,8 @@ private void InitializeWindow()
             string titleIdSection = string.IsNullOrWhiteSpace(activeProcess.ProgramIdText) ? string.Empty : $" ({activeProcess.ProgramIdText.ToUpper()})";
             string titleArchSection = activeProcess.Is64Bit ? " (64-bit)" : " (32-bit)";
 
-            WindowHandle = SDL_CreateWindow($"Ryujinx {Program.Version}{titleNameSection}{titleVersionSection}{titleIdSection}{titleArchSection}", SDL_WINDOWPOS_UNDEFINED, SDL_WINDOWPOS_UNDEFINED, DefaultWidth, DefaultHeight, DefaultFlags | GetWindowFlags());
+            SDL_WindowFlags fullscreenFlag =  IsFullscreen ? SDL_WindowFlags.SDL_WINDOW_FULLSCREEN_DESKTOP : 0;
```
Comment (AcK77) [inline src/Ryujinx.Headless.SDL2/WindowBase.cs:160]:
```suggestion
            SDL_WindowFlags fullscreenFlag =  IsFullscreen ? SDL_WindowFlags.SDL_WINDOW_FULLSCREEN_DESKTOP : 0;

```

**Review by AcK77 — APPROVED** [review]

**Comment by siegmund-heiss-ich** [comment]
Currently macOS crashes if you try to start headless in fullscreen mode.
This needs to be resolved before merge.

No useful information in ryujinx.log
macOS crash report:
```
-------------------------------------
Translated Report (Full Report Below)
-------------------------------------

Process:               Ryujinx.Headless.SDL2 [2927]
Path:                  /Users/USER/Desktop/*/Ryujinx.Headless.SDL2
Identifier:            Ryujinx.Headless.SDL2
Version:               ???
Code Type:             ARM-64 (Native)
Parent Process:        zsh [1178]
Responsible:           iTerm2 [1174]
User ID:               501

Date/Time:             2023-06-29 17:09:35.9143 +0200
OS Version:            macOS 13.4.1 (22F82)
Report Version:        12
Anonymous UUID:        6DF772F6-6CCF-2D06-1510-EBE77F2E1FEC


Time Awake Since Boot: 1000 seconds

System Integrity Protection: enabled

Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000

Application Specific Information:
abort() called


Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	       0x183224724 __pthread_kill + 8
1   libsystem_pthread.dylib       	       0x18325bc28 pthread_kill + 288
2   libsystem_c.dylib             	       0x183169ae8 abort + 180
3   libcoreclr.dylib              	       0x102942688 PROCAbort + 56
4   libcoreclr.dylib              	       0x1029425a4 PROCEndProcess(void*, unsigned int, int) + 384
5   libcoreclr.dylib              	       0x102b5a0e4 UnwindManagedExceptionPass1(PAL_SEHException&, _CONTEXT*) + 972
6   libcoreclr.dylib              	       0x102b5a12c DispatchManagedException(PAL_SEHException&, bool) + 68
7   libcoreclr.dylib              	       0x102a3e54c VSD_ResolveWorker + 1552
8   libcoreclr.dylib              	       0x102be3a4c ResolveWorkerAsmStub + 88
9   ???                           	       0x283698620 ???
10  ???                           	       0x283698620 ???
```

**Comment by TSRBerry** [comment]
My test results:

<details>
<summary>Linux x64: OpenGL</summary>

Crash:
```
System.InvalidOperationException: SDL_GL function returned an error: No OpenGL context has been made current
   at Ryujinx.Headless.SDL2.OpenGL.OpenGLWindow.CheckResult(Int32 result) in /src/Ryujinx/src/Ryujinx.Headless.SDL2/OpenGL/OpenGLWindow.cs:line 18
   at Ryujinx.Headless.SDL2.OpenGL.OpenGLWindow.InitializeWindowRenderer() in /src/Ryujinx/src/Ryujinx.Headless.SDL2/OpenGL/OpenGLWindow.cs:line 131
   at Ryujinx.Headless.SDL2.WindowBase.Render() in /src/Ryujinx/src/Ryujinx.Headless.SDL2/WindowBase.cs:line 231
```
</details> 

<details>
<summary>Linux x64: Vulkan</summary>

Works fine!
</details> 

<details>
<summary>macOS arm64: Vulkan</summary>

Works fine!
</details>


#### Notes

- How am I supposed to stop emulation?
- I noticed touchscreen input is currently not recognized properly, but this might not be an issue with your changes, as I think I'm encountering the same issues on master right now.

**Comment by siegmund-heiss-ich** [comment]
>     * How am I supposed to stop emulation?
> 
>     * I noticed touchscreen input is currently not recognized properly, but this might not be an issue with your changes, as I think I'm encountering the same issues on master right now.

With Control + C. I suppose users who use it in CLI would know that. Tbh I don't know which other use cases there would be. Unless you use some game frontend ui thing to control it. And in this case there is hopefully a way to stop emulation directly from this app. 
The use of a shortcut could be implemented here. But I don't think it's necessary.

Isn't touchscreen implemented with ava?

**Review by marysaka — APPROVED** [review]

**Review by marysaka — APPROVED** [review]

**Pull request merged**